File: android/app/src/main/java/com/nsfy/app/data/model/NotificationPolicy.kt
package com.nsfy.app.data.model

const val KEY_DO_NOT_DISTURB = "do_not_disturb"
const val KEY_DND_PRIORITIES = "dnd_allowed_priorities"
const val KEY_NOTIFICATION_MODE = "notification_mode"

enum class NotificationMode(val value: String) {
    Silent("silent"),
    System("system"),
    Temporary("temporary"),
    Persistent("persistent");

    companion object {
        fun from(value: String?): NotificationMode = entries.firstOrNull {
            it.value == value
        } ?: System
    }
}

fun shouldPresentNotification(
    message: NsfyMessage,
    doNotDisturb: Boolean,
    allowedPriorities: Set<Int>,
): Boolean = message.popup && (
    !doNotDisturb || message.bypassDnd || message.priority in allowedPriorities
)

File: desktop/src/lib/notification-policy.ts
export interface DeliveryPolicy {
  priority: number;
  popup: boolean;
  bypassDnd: boolean;
}

export type Presentation = 'none' | 'main' | 'system' | 'temporary' | 'persistent';

export function selectPresentation(
  mode: 'silent' | 'system' | 'temporary' | 'persistent',
  windowBehavior: 'popup' | 'resident',
  systemPermission: boolean,
): Presentation {
  if (mode === 'silent') return 'none';
  if (windowBehavior === 'popup') return 'main';
  if (mode === 'system') return systemPermission ? 'system' : 'none';
  return mode;
}

export function shouldPresentNotification(
  message: DeliveryPolicy,
  doNotDisturb: boolean,
  allowedPriorities: number[],
): boolean {
  if (!message.popup) return false;
  return !doNotDisturb
    || message.bypassDnd
    || allowedPriorities.includes(message.priority);
}
